components/post_view/reaction/reaction.jsx

```diff
@@ -83,15 +85,155 @@ export default class Reaction extends React.PureComponent {
         sortedUsers: PropTypes.object.isRequired,
     }
 
-    handleAddReaction = (e) => {
-        e.preventDefault();
-        const {actions, post, emojiName} = this.props;
-        actions.addReaction(post.id, emojiName);
+    constructor(props) {
+        super(props);
+
+        const {reactionCount, canRemoveReaction, canAddReaction} = this.props;
+        const {currentUserReacted} = this.props.sortedUsers;
+
+        this.state = {
+            canAddReaction,
+            canRemoveReaction,
+        };
+
+        if (currentUserReacted) {
+            this.state = {
+                ...this.state,
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction--reacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount,
+                unreactedNumber: reactionCount - 1,
+            };
+        } else {
+            this.state = {
+                ...this.state,
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction--unreacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount + 1,
+                unreactedNumber: reactionCount,
+            };
+        }
+
+        if (!canAddReaction || !canRemoveReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction--read-only',
+            };
+        }
+
+        this.reactionButtonRef = React.createRef();
+        this.reactionCountRef = React.createRef();
+        this.reactedNumeralRef = React.createRef();
+    }
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        // reaction count has changed, but not by current user
+        if (nextProps.reactionCount !== prevState.displayNumber && nextProps.sortedUsers.currentUserReacted === prevState.userReacted) {
```

Comment:
What if multiple of these props change at the same time? This logic seems incorrect if, for example, the permissions and reaction count change at the same time.

Reply:
I've tested this thoroughly with several concurrent users so I don't think there's a problem, especially now that the canAdd/canRemove permissions aren't being handled here.

Essentially if the reaction of this user has NOT changed, but the reaction count has changed, we update the numbers (either up or down). Any other changes from props will be ignored.